src/main/java/com/amazon/opendistroforelasticsearch/ad/rest/handler/ValidateAnomalyDetectorActionHandler.java

```diff
@@ -0,0 +1,590 @@
+    }
+
+    private void queryFilterValidation() {
+        long[] startEnd = startEndTimeRangeWithIntervals(NUM_OF_INTERVALS_CHECKED_FILTER);
+        long dataEndTime = startEnd[1];
+        long dataStartTime = startEnd[0];
+        RangeQueryBuilder rangeQuery = new RangeQueryBuilder(anomalyDetector.getTimeField())
+            .from(dataStartTime)
+            .to(dataEndTime)
+            .format("epoch_millis");
+        BoolQueryBuilder internalFilterQuery = QueryBuilders.boolQuery().must(rangeQuery).must(anomalyDetector.getFilterQuery());
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+            .query(internalFilterQuery)
+            .size(1)
+            .terminateAfter(1)
+            .timeout(requestTimeout);
+        SearchRequest searchRequest = new SearchRequest(anomalyDetector.getIndices().get(0)).source(searchSourceBuilder);
+        client
+            .search(
+                searchRequest,
+                ActionListener.wrap(searchResponse -> onQueryFilterSearch(searchResponse), exception -> { onFailure(exception); })
+            );
+    }
+
+    private void onQueryFilterSearch(SearchResponse response) throws IOException {
+        if (response.getHits().getTotalHits().value <= 0) {
+            List<String> filterError = new ArrayList<>();
+            filterError
+                .add(
+                    "query filter is potentially wrong as no hits were found at all or no historical data in last "
+                        + NUM_OF_INTERVALS_CHECKED_FILTER
+                        + " intervals"
+                );
+            suggestedChangesMap.put("filter_query", filterError);
+            validateAnomalyDetectorResponse();
+        } else {
+            long timestamp = (long) response.getHits().getHits()[0].getSourceAsMap().get("timestamp");
+            featureQueryValidation(timestamp);
+        }
+    }
+
+    private void featureQueryValidation(long startTimeStamp) throws IOException {
+        long delayMillis = Optional
+            .ofNullable((IntervalTimeConfiguration) anomalyDetector.getWindowDelay())
+            .map(t -> t.toDuration().toMillis())
+            .orElse(0L);
+        long[] startEnd = startEndTimeRangeWithIntervals(NUM_OF_INTERVALS_CHECKED);
+        long dataEndTime = startEnd[1];
+        long dataStartTime = startEnd[0];
+        if (startEnd[0] > startTimeStamp) {
+            dataStartTime = startTimeStamp;
+        }
+        IntervalTimeConfiguration searchRange = new IntervalTimeConfiguration(FEATURE_VALIDATION_TIME_BACK_MINUTES, ChronoUnit.MINUTES);
+        long searchRangeTime = Optional.ofNullable(searchRange).map(t -> t.toDuration().toMillis()).orElse(0L);
+        long startTimeWithSetTime = startEnd[1] - (searchRangeTime - delayMillis);
+        // Make sure start time includes timestamp seen by filter query check
+        if (startEnd[0] > startTimeWithSetTime) {
+            dataStartTime = startTimeWithSetTime;
+        }
```

Comment:
I would suggest you move such lines into a separate method like `getFeatureQueryValidationDateRange`

Reply:
done